File: http-core/src/main/scala/org/apache/pekko/http/scaladsl/model/FormData.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka.
 */

/*
 * Copyright (C) 2009-2022 Lightbend Inc. <https://www.lightbend.com>
 */

package org.apache.pekko.http.scaladsl.model

import org.apache.pekko
import pekko.http.impl.model.parser.CharacterClasses
import pekko.http.impl.util.StringRendering
import pekko.http.scaladsl.model.MediaTypes._

/**
 * Simple model for `application/x-www-form-urlencoded` form data.
 */
final case class FormData(fields: Uri.Query) {
  def toEntity: pekko.http.scaladsl.model.RequestEntity =
    toEntityWithCharset(`application/x-www-form-urlencoded`.charset)

  private def toEntityWithCharset(charset: HttpCharset): pekko.http.scaladsl.model.RequestEntity = {
    val render: StringRendering =
      UriRendering.renderQuery(new StringRendering, this.fields, charset.nioCharset, CharacterClasses.unreserved)
    HttpEntity(`application/x-www-form-urlencoded`, render.get)
  }
}

object FormData {
  val Empty = FormData(Uri.Query.Empty)

  def apply(fields: Map[String, String]): FormData =
    if (fields.isEmpty) Empty else FormData(Uri.Query(fields))

  def apply(fields: (String, String)*): FormData =
    if (fields.isEmpty) Empty else FormData(Uri.Query(fields: _*))
}

File: http-core/src/main/java/org/apache/pekko/http/javadsl/model/FormData.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * license agreements; and to You under the Apache License, version 2.0:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * This file is part of the Apache Pekko project, which was derived from Akka.
 */

/*
 * Copyright (C) 2009-2022 Lightbend Inc. <https://www.lightbend.com>
 */

package org.apache.pekko.http.javadsl.model;

import java.util.Map;

import org.apache.pekko.japi.Pair;

/** Simple model for `application/x-www-form-urlencoded` form data. */
public final class FormData {

  private final Query fields;

  public FormData(Query fields) {
    this.fields = fields;
  }

  /**
   * Returns the fields of this FormData instance.
   *
   * @return the fields
   * @since 2.0.0
   */
  public Query getFields() {
    return fields;
  }

  /** Converts this FormData to a RequestEntity using UTF8 encoding. */
  public RequestEntity toEntity() {
    return HttpEntities.create(
        ContentTypes.APPLICATION_X_WWW_FORM_URLENCODED, fields.render(HttpCharsets.UTF_8));
  }

  /** Returns empty FormData. */
  public static final FormData EMPTY = new FormData(Query.EMPTY);

  /** Creates the FormData from the given parameters. */
  @SafeVarargs
  public static FormData create(Pair<String, String>... params) {
    return new FormData(Query.create(params));
  }

  /** Creates the FormData from the given parameters. */
  public static FormData create(Map<String, String> params) {
    return new FormData(Query.create(params));
  }

  /** Creates a FormData from the given parameters. */
  public static FormData create(Iterable<Pair<String, String>> params) {
    return new FormData(Query.create(params));
  }
}
